documentation/PreferencesServer.md

```diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+with the contents of the system error:
+
+```json
+{
+    "isError": true,
+    "errorCode": "GPII_ERR_APPLICABLE_MESSAGE",
+    "message": "System error described here."
+}
+```
+
+### PUT /prefssafe/:prefsSafeId
+
+Updates an existing preferences safe in the database, using the full preferences safe format. Will return the
+updated safe, which should include an updated `timestampUpdated` field.
+
+#### Example PUT
+
+URL: `http://preferences.gpii.net/prefssafe/prefsSafe-alice`
+
+Example PUT body:
+
+```json
+{
+    "id": "prefsSafe-alice",
+    "type": "prefsSafe",
+    "schemaVersion": "0.1",
+    "prefsSafeType": "user",
+    "name": "alice",
+    "email": null,
+    "preferences": {
+        "flat": {
+            "name": "Alice",
+            "contexts": {
+                "gpii-default": {
+                    "name": "Default preferences",
+                    "preferences": {
+                        "http://registry.gpii.net/applications/com.microsoft.windows.onscreenKeyboard": {}
+                    }
+                }
+            }
+        }
+    },
+    "timestampCreated": "2019-01-14T23:48:03.221Z",
+    "timestampUpdated": null
+}
+```
+
+On a successfull save, the same payload from the PUT operation will be returned, with an updated
+`timestampUpdated` field with the time of save. In the event of a failed save an error payload with a
+suitable message will be returned.
+
+```json
+{
+    "isError": true,
+    "errorCode": "APPROPRIATE_ERROR_CODE",
+    "message": "System error described here."
+}
+```
+
+### GET /prefssafes
+
+Returns a list of preferences safes, including only basic information about each one. Ideal for building a table
+or listing of preferences safes. Each item in the list representing a preferences safe will include `id`, `name`,
+`email`, `created`, and `updated`. This endpoint will likely have more options in the future, such as sorting,
+paging, etc.
+
+#### Example GET
+
+URL: `http://preferences.gpii.net/prefssafes`
+
+Example return payload:
+
+```json
+{
+    "total_rows": 2,
+    "offset": 0,
+    "rows": [
+        {
+            "name": "Alice",
+            "email": "alice@gpii.net",
+            "created": "2017-12-14T19:55:11.640Z",
+            "updated": null,
+            "id": "prefsSafe-1"
+        },
+        {
+            "name": null,
+            "email": null,
+            "created": "2017-12-14T19:55:11.640Z",
+            "updated": null,
+            "id": "prefsSafe-2"
+        }
+    ]
+}
+```
+
+There should never be an error payload for this endpoint. In the situation where there are no preferences
+safes stored in the system it will merely contain zero rows.
+
+### GET /prefssafe-keys/:prefsSafeId
+
+This will return the associated keys and credentials documents for a given preferences safe, in a `rows` field.
+Also included is a `total_rows` and `offset` field, but do note that the `total_rows` field is not accurate as

```

Comment:
What's `offset` field for? Is there a reason that `total_rows` value is left inaccurate?

Reply:
`offset` is a standard couch field that indicated how far from the beginning of the dataset your first item is.  This is allow for paging if you are only getting a subset of the return data.  

I can't remember why `offset` doens't actually work the way you would expect in couch. I think @the-t-in-rtf explained it to me once.  It might be a good thing for us to address as a whole project wise as we'll have more and more endpionts returning arrays.  For now, I've just documented it as such.

Reply:
Just out of curiosity, I did some search and experiment to see what's going on with `total_rows`. It turns out this field seems returning the correct number it's supposed to return.

[The CouchDB doc](http://docs.couchdb.org/en/stable/api/ddoc/views.html#db-design-design-doc-view-view-name) says `totoal_rows` holds "Number of documents in the database/view". This endpoint uses the view "findRelatedDocsForPrefsSafe" so `total_rows` should have the value of total number of documents in this view.

If you do a CouchDB query on this view without providing a specific `key` value: `/gpii/_design/views/_view/findRelatedDocsForPrefsSafe?include_docs=true`, the query returns all documents in this view, whose total matches the value of `total_rows` when this view is queried with a specific key.

If you could double confirm on this, the statement of "the `total_rows` field is not accurate" can be removed from here.